lib/mayaUsd/utils/targetLayer.cpp

```diff
@@ -28,6 +30,54 @@
 #include <maya/MObject.h>
 #include <maya/MString.h>
 
+namespace {
+
+const char kTargetLayerAttrName[] = "usdStageTargetLayer";
+const char kTargetLayerPrimPathAttrName[] = "usdStageTargetLayerPrimPath";
+
+/// \brief Find prim node that matches given layer id
+PXR_NS::PcpNodeRef
+findPrimNode(const PXR_NS::PcpNodeRef& primNode, const std::string& targetLayerId)
+{
+    // We need to store and check current depth level first (BFS), this is to
+    // preserve the opinions from strong to week
+    std::vector<PXR_NS::PcpNodeRef> childNodes;
+    TF_FOR_ALL(child, primNode.GetChildrenRange())
+    {
+        // The prim node can't have a variant.
+        if (child->GetPath().ContainsPrimVariantSelection()) {
+            continue;
+        }
+        // Confirm that the primNode is a direct contributor to the root prim.
+        bool result = child->IsRootNode();
+        // Confirm we are looking at one of the composition arcs that could lead us to a prim
+        // instance and not some other contributor like a specializes or such.
+        if (!result) {
+            auto arcType = child->GetArcType();
+            result = (arcType == PcpArcTypeReference) || (arcType == PcpArcTypePayload)
+                || (arcType == PcpArcTypeVariant);
+        }
```

Comment:
If I understand the code above, you search for either a root node or an ref/payload/variant arc?

Reply:
Yes, this logic has been used in our studio for a few years and haven't received anything from artists complaining returning unexpected layer, but I am happy to review the code if other studios encounter any problem with that.